ERCS/erc-7920.md

```diff
@@ -0,0 +1,299 @@
+
+Current solutions have significant drawbacks:
+
+- **Pre-approving [ERC-20](./eip-20.md) allowance:** spend creates security vulnerabilities
+- **Merging multiple messages into a single message:** prevents independent verifiability. Each message cannot be verified without knowledge of the entire batch
+- **Separate signature requests:** creates friction in the user experience
+
+## Specification
+
+### Overview
+
+The composite signature scheme uses a Merkle tree to hash multiple typed-data data messages together under a single root. The user signs only the Merkle root. The process is described below.
+
+### Generating a Composite Signature
+
+1. For a set of messages `[m₁, m₂, ..., mₙ]`, encode each using EIP-712's `encode` and compute its hash:
+
+   ```
+   hashₙ = keccak256(encode(mₙ))
+   ```
+
+2. Use these message hashes as leaf nodes in a Merkle tree and compute a `marketRoot`
+
+3. Sign the merkle root.
+
+   ```
+   signature = sign(marketRoot)
+   ```
+
+### Verification Process
+
+To verify that an individual message `mₓ` was included in a composite signature:
+
+1. Verify the signature on the `merkleRoot`:
+
+   ```
+   recoveredSigner = ecrecover(merkleRoot, signature)
+   isValidSignature = (recoveredSigner == expectedSigner)
+   ```
+
+2. Compute the leaf node for message `mₓ` and verify it's path to the Merkle root, using the proof:
+   ```
+   leaf = keccak256(encode(mₓ))
+   isValidProof = _verifyMerkleProof(leaf, merkleProof, merkleRoot)
+   ```
+
+The message is verified if and only if (1) and (2) succeed.
+
+```
+isVerified = isValidSignature && isValidProof
+```
+
+#### Solidity Example
+
+Snippet below verifies a composite signature on-chain. The entrypoint, `placeOrder()` is called by a paymaster to sponsor an operation. It verifies the composite signature using the process outlined above.
+
+```solidity
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.20;
+
+error Unauthorized();
+
+contract ExampleVerifier {
+    bytes32 public immutable DOMAIN_SEPARATOR;
+    bytes32 private constant MESSAGE_TYPEHASH =
+        keccak256("PlaceOrder(bytes32 orderId, address user)");
+
+    constructor() {
+        DOMAIN_SEPARATOR = keccak256(
+            abi.encode(
+                keccak256(
+                    "EIP712Domain(string name,string version,uint256 chainId,address verifyingContract)"
+                ),
+                keccak256(bytes("MyApp")),
+                keccak256(bytes("1.0.0")),
+                block.chainid,
+                address(this)
+            )
+        );
+    }
+
+    function placeOrder(
+        bytes32 orderId,
+        address user,
+        bytes calldata signature,
+        bytes32 merkleRoot,
+        bytes32[] calldata proof
+    ) public {
+        bytes32 message = keccak256(
+            abi.encodePacked(
```

Comment:
```suggestion
            abi.encode(
```

Attacker could potentially leverage that `abi.encodePacked` to create hash collisions. Is it intentional?

abi.encode adds padding to maintain 32-byte alignment for each element which is safer. And you were using it in line 92

Reply:
Overlooked. Will fix